rust/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,501 @@
+use super::emitter::{AtomIndex, EmitError};
+use crate::ast_emitter::AstEmitter;
+
+#[derive(Debug, PartialEq)]
+enum SetOpReferenceKind {
+    GlobalName(AtomIndex),
+    #[allow(dead_code)]
+    Prop(AtomIndex),
+    #[allow(dead_code)]
+    Elem,
+}
+
+// See SetOpReferenceEmitter.
+// Use struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct SetOpReference {
+    kind: SetOpReferenceKind,
+}
+impl SetOpReference {
+    fn new(kind: SetOpReferenceKind) -> Self {
+        Self { kind }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum CallOpKind {
+    Normal,
+    // FIXME: Support eval, call, apply etc.
+}
+
+// See CallOpReferenceEmitter.
+// Use struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct CallOpReference {
+    kind: CallOpKind,
+}
+impl CallOpReference {
+    fn new(kind: CallOpKind) -> Self {
+        Self { kind }
+    }
+}
+
+// Struct for emitting bytecode for get operation.
+//
+// Usage:
+//
+//   `name;`
+//      GetEmitter::emit_name(self, name);
+//
+//   `obj.key;`
+//      GetEmitter::emit_prop(self, |emitter| {
+//          emitter.emit_expression(obj)
+//      }, key)?;
+//
+//   `obj[key];`
+//      GetEmitter::emit_elem(self, |emitter| {
+//          emitter.emit_expression(obj)
+//      }, |emitter| {
+//          emitter.emit_expression(key)
+//      })?;
+//
+pub struct GetEmitter {}
+impl GetEmitter {
+    pub fn emit_name(emitter: &mut AstEmitter, name: &str) {
+        let name_index = emitter.emit.get_atom_index(name);
+
+        //              [stack]
+
+        // FIXME: Support non-global case.
+        emitter.emit.get_g_name(name_index);
+        //              [stack] VAL
+    }
+
+    pub fn emit_prop<F>(
+        emitter: &mut AstEmitter,
+        obj_emitter: F,
+        key: &str,
+    ) -> Result<(), EmitError>
+    where
+        F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    {
+        let key_index = emitter.emit.get_atom_index(key);
+
+        //              [stack]
+
+        obj_emitter(emitter)?;
+        //              [stack] OBJ
+
+        emitter.emit.get_prop(key_index);
+        //              [stack] VAL
+
+        Ok(())
+    }
+
+    pub fn emit_prop_super<F>(
+        emitter: &mut AstEmitter,
+        this_emitter: F,
+        key: &str,
+    ) -> Result<(), EmitError>
+    where
+        F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    {
+        let key_index = emitter.emit.get_atom_index(key);
+
+        //              [stack]
+
+        this_emitter(emitter)?;
+        //              [stack] THIS
+
+        emitter.emit.callee();
+        //              [stack] THIS CALLEE
+
+        emitter.emit.super_base();
+        //              [stack] THIS OBJ
+
+        emitter.emit.get_prop_super(key_index);
+        //              [stack] VAL
+
+        Ok(())
+    }
+
+    pub fn emit_elem<F1, F2>(
+        emitter: &mut AstEmitter,
+        obj_emitter: F1,
```

Comment:
Seems like it would be quite easy to get this order wrong.

Reply:
yeah, it's one of the issues needs to be solved eventually.
Rust doesn't have named parameter, and at this point `obj_emitter` and `key_emitter` will have exactly same type, even with stack in/out type added...

Reply:
A way of solving the order issue without repeating type name (so, not using struct for parameter) would be, split the helper struct for each method, and make the helper struct constructible, so, something like:

```
struct GetElemEmitter<F1, F2>
where
    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
{
    obj: F1,
    key: F2,
}

impl<F1, F2> GetElemEmitter<F1, F2>
where
  F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
  F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
{
    fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
        //              [stack]

        self.obj(emitter)?;
        //              [stack] OBJ

        self.key(emitter)?;
        //              [stack] OBJ KEY

        emitter.emit.get_elem();
        //              [stack] VAL

        Ok(())
    }
}

// caller
GetElemEmitter {
    obj: |emitter| {
        emitter.emit_expression(obj)
    },
    key: |emitter| {
        emitter.emit_expression(key)
    },
}.emit(self)?;
```

(not yet tested)

Reply:
Here's the patch that uses the above way: https://github.com/arai-a/jsparagus/commit/7e743b87bc8fac5a104d7b9dfe13a8e3525f0daa